### DISCUSSION TITLE
Read out font information

### DISCUSSION THREAD

**Discussion by k-ljung** [RudolphRiedel/FT800-FT813#77] [General]
I am in the process of replacing Riverdi's display library with yours in one of my projects, but have a function where I read out information about fonts that I am unable to translate into your library. Among other things, I read out the height in pixels of the font.

This struct is copied from Riverdi's library:
```
/* FT81x and FT80x font table structure */
/* Font table address in ROM can be found by reading the address from 0xFFFFC location. */
/* 16 font tables are present at the address read from location 0xFFFFC */
typedef struct Gpu_Fonts
{
    /* All the values are in bytes */
    /* Width of each character font from 0 to 127 */
    uint8_t    FontWidth[GPU_NUMCHAR_PERFONT];
    /* Bitmap format of font wrt bitmap formats supported by FT800 - L1, L4, L8 */
    uint32_t    FontBitmapFormat;
    /* Font line stride in FT800 ROM */
    uint32_t    FontLineStride;
    /* Font width in pixels */
    uint32_t    FontWidthInPixels;
    /* Font height in pixels */
    uint32_t    FontHeightInPixels;
    /* Pointer to font graphics raw data */
    uint32_t    PointerToFontGraphicsData;
}Gpu_Fonts_t;
```
I have my method where I read out the information and creating the `Gpu_Fonts` struct
```
Gpu_Fonts_t App_Read_FontMetric(int16_t font)
{
    uint8_t font_data[148];    
    uint32_t adr_width = ROM_ROOT_FONT + (148* (font-16));
    
    Gpu_Hal_RdMem(phost, adr_width, font_data, 148);

    Gpu_Fonts fontMetric;

    for (size_t i = 0; i < 128; i++)
    {
        fontMetric.FontWidth[i] = font_data[1];
    }
    
    fontMetric.FontBitmapFormat = font_data[128] | (font_data[129] << 8) | (font_data[130] << 16) | (font_data[131] << 24);
    fontMetric.FontLineStride = font_data[132] | (font_data[133] << 8) | (font_data[134] << 16) | (font_data[135] << 24);
    fontMetric.FontWidthInPixels = font_data[136] | (font_data[137] << 8) | (font_data[138] << 16) | (font_data[139] << 24);
    fontMetric.FontHeightInPixels = font_data[140] | (font_data[141] << 8) | (font_data[142] << 16) | (font_data[143] << 24);
    fontMetric.PointerToFontGraphicsData = font_data[144] | (font_data[145] << 8) | (font_data[146] << 16) | (font_data[147] << 24);
    
    return fontMetric;
}
```

It's specific `Gpu_Hal_RdMem(phost, adr_width, font_data, 148);` method that I don't understand how to translate to your library.  In Riverdi's library it does the following:
```
/*
 * Gpu_Hal_RdMem()
 */
void
Gpu_Hal_RdMem (Gpu_Hal_Context_t  *host,
               uint32_t            addr,
               uint8_t            *buffer,
               uint32_t            length)
{
  Gpu_Hal_StartTransfer(host,GPU_READ,addr);

  while (length--)
    {
      *buffer = Gpu_Hal_Transfer8(host,0);
      buffer++;
    }

  Gpu_Hal_EndTransfer(host);
}
```
Any ideas how I can read it out of memory and populate my `font_data` variable?

**Comment by RudolphRiedel** [comment]
I really had to check, I never implemented a function that reads from EVE to a buffer on the controller. :-)
Well, you can just loop thru EVE_memRead8() calls.

This might work as well:
`
/* Helper function, read a block of memory from EVE to the SRAM of the host controller */
void EVE_memRead_sram_buffer(uint32_t ft_address, const uint8_t *p_data, uint32_t len)
{
    EVE_cs_set();
    spi_transmit_32(((ft_address >> 16U) & 0x0000007fUL) + (ft_address & 0x0000ff00UL) + ((ft_address & 0x000000ffUL) << 16U));

    for (uint32_t count = 0U; count < length; count++)
    {
        p_data[count] = spi_receive(0U); /* read data byte by sending another dummy byte */
    }

    EVE_cs_clear();
}
`

But I did not even compile it, yet. :-)

Edit: and I probably should add a check for NULL.

**Reply by k-ljung** [reply]
Thanks for your quick answer.

There was a small typo but it led me in the right direction, this method works.
```
      /* Helper function, read a block of memory from EVE to the SRAM of the host controller */
        void EVE_memRead_sram_buffer(uint32_t ft_address, uint8_t *buffer, uint32_t length)
        {
            EVE_cs_set();
            spi_transmit_32(((ft_address >> 16U) & 0x0000007fUL) + (ft_address & 0x0000ff00UL) + ((ft_address & 0x000000ffUL) << 16U));

            while (length--)
            {
                *buffer = spi_receive(0U);
                buffer++;
            }

            EVE_cs_clear();
        }
```

**Reply by RudolphRiedel** [reply]
I am still trying to figure out why the data that is beeing read does not seem to be correct. :-)

**Reply by RudolphRiedel** [reply]
I am using a BT815, it does not have the table at EVE_ROM_FONT.
One needs to read EVE_ROM_FONTROOT to receive the address of the table.
And EVE_ROM_FONTROOT was named EVE_ROM_FONT_ADDR so I fixed this as well.  

uint32_t fontbase = EVE_memRead32(EVE_ROM_FONTROOT );  
EVE_memRead_sram_buffer(fontbase + (148 * (28 - 16)), (uint8_t *) &rom_font, 148); /* read data for ROM font 28 */  

The data however does not match the table in the datasheet.
For font 28 I get:
rom_font.format: 2 -> L4
rom_font.line_stride: 9
rom_font.pixel_width: 18
rom_font.pixel_height: 25

And the datasheet has this values: L4 / 9 / 19 / 25

Well, close enough I guess and seems to be working. :-)
